### PR TITLE
lru cache + some pytest setup

### PR DESCRIPTION
**Pull request by lisss**

### CODE REVIEW

File: data_structures/cache.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+class DLNode:
+    def __init__(self, key, data):
+        self.prev = None
+        self.next = None
+        self.key = key
+        self.data = data
+
+
+class DLList:
```
Comment (lisss) [inline data_structures/cache.py:9]: i have no idea how to do it properly without an additional class... what the bottlenecks are with it?

Comment (RayanRal) [reply data_structures/cache.py:9]: no worries, we can just put those utility methods in there, and that's totally ok.

File: data_structures/cache.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            self.head = None
+            return key
+        else:
+            key = self.tail.key
+            self.tail.prev.next = None
+            self.tail = self.tail.prev
+            return key
+
+
+def shift_node(node, llist):
```
Comment (RayanRal) [inline data_structures/cache.py:37]: yep, that should work.

**Review by RayanRal — APPROVED** [review]
yep, conceptually correct, maybe some checks or pointers can be missing, but definitely nothing major

**Pull request merged**